pkg/bpf/exporter.go

```diff
@@ -82,285 +69,220 @@ func NewExporter() (Exporter, error) {
 	return e, err
 }
 
-type perfCounter struct {
-	EvType   int
-	EvConfig int
-}
-
 func (e *exporter) SupportedMetrics() SupportedMetrics {
 	return SupportedMetrics{
 		HardwareCounters: e.enabledHardwareCounters.Clone(),
 		SoftwareCounters: e.enabledSoftwareCounters.Clone(),
 	}
 }
 
-func getLibbpfObjectFilePath(byteOrder binary.ByteOrder) (string, error) {
-	var endianness string
-	if byteOrder == binary.LittleEndian {
-		endianness = "bpfel"
-	} else if byteOrder == binary.BigEndian {
-		endianness = "bpfeb"
-	}
-	filename := fmt.Sprintf(objectFilename, endianness)
-	bpfassetsPath := fmt.Sprintf("%s/%s", bpfAssestsLocation, filename)
-	_, err := os.Stat(bpfassetsPath)
-	if err != nil {
-		// attempt to find the bpf assets in the same directory as the binary
-		// this is useful for running locally
-		var matches []string
-		err = filepath.Walk(".", func(path string, info os.FileInfo, err error) error {
-			if info.Name() == filename {
-				matches = append(matches, path)
-				return filepath.SkipAll
-			}
-			return nil
-		})
-		if err != nil {
-			return "", fmt.Errorf("failed to find bpf object file: %w", err)
-		}
-		if len(matches) < 1 {
-			return "", fmt.Errorf("failed to find bpf object file: no matches found")
-		}
-		klog.Infof("found bpf object file: %s", matches[0])
-		return matches[0], nil
-	}
-	return bpfassetsPath, nil
-}
-
 func (e *exporter) attach() error {
-	libbpfObjectFilePath, err := getLibbpfObjectFilePath(e.byteOrder)
-	if err != nil {
-		return fmt.Errorf("failed to load module: %w", err)
+	// Remove resource limits for kernels <5.11.
+	if err := rlimit.RemoveMemlock(); err != nil {
+		return fmt.Errorf("error removing memlock: %v", err)
 	}
 
-	e.module, err = bpf.NewModuleFromFile(libbpfObjectFilePath)
+	// Load eBPF Specs
+	specs, err := loadKepler()
 	if err != nil {
-		return fmt.Errorf("failed to load eBPF module from libbpf object: %w", err)
+		return fmt.Errorf("error loading eBPF specs: %v", err)
 	}
 
-	// resize array entries
-	klog.Infof("%d CPU cores detected. Resizing eBPF Perf Event Arrays", e.cpuCores)
-	toResize := []string{
-		"cpu_cycles_event_reader", "cpu_instructions_event_reader", "cache_miss_event_reader", "task_clock_ms_event_reader",
-		"cpu_cycles", "cpu_instructions", "cache_miss", "cpu_freq_array", "task_clock",
-	}
-	for _, arrayName := range toResize {
-		if err = resizeArrayEntries(e.module, arrayName, e.cpuCores); err != nil {
-			return fmt.Errorf("failed to resize array %s: %w", arrayName, err)
+	// Adjust map sizes to the number of available CPUs
+	numCPU := getCPUCores()
+	klog.Infof("Number of CPUs: %d", numCPU)
+	for _, m := range specs.Maps {
+		// Only resize maps that have a MaxEntries of NUM_CPUS constant
+		if m.MaxEntries == 128 {
```

Comment:
can we check condition by map name prefix if not by map name? the size comparison does not reflect why are we resizing the map as per number of cpus.

Reply:
Well, 128 is indeed the magic `NUM_CPUS` constant that we define on the eBPF side.
If you use `NUM_CPUS` as your max-entries value, the contract is that userspace will resize your map to the number of CPUs.

I prefer this to keeping a list of map names and/or prefixes around